k2/torch/csrc/fsa_class.cu

```diff
@@ -0,0 +1,476 @@
+
+  std::vector<Ragged<int32_t>> raggeds;
+  for (const auto &attr : ragged_tensor_attrs) {
+    // skip this attribute, as it is not included in all source Fsas.
+    if (attr.second != srcs.size()) continue;
+    for (const auto &fsa : srcs) {
+      auto iter = fsa.ragged_tensor_attrs.find(attr.first);
+      K2_CHECK(iter != fsa.ragged_tensor_attrs.end());
+      raggeds.emplace_back(iter->second);
+    }
+    auto value = Cat<int32_t>(/*axis*/ 0, raggeds.size(), raggeds.data(),
+                              /*merge_map*/ nullptr);
+    SetRaggedTensorAttr(attr.first, value);
+  }
+}
+
+void FsaClass::SetScores(torch::Tensor scores) {
+  K2_CHECK_EQ(scores.numel(), fsa.NumElements());
+  K2_CHECK_EQ(scores.scalar_type(), torch::kFloat32);
+  Scores().copy_(scores.detach());
+}
+
+torch::Tensor FsaClass::Scores() {
+  auto device = DeviceFromContext(fsa.Context());
+  auto scalar_type = caffe2::TypeMeta::Make<float>();
+
+  // an Arc has 4 members
+  static_assert(sizeof(Arc) == 4 * sizeof(int32_t), "");
+
+  std::vector<int64_t> sizes = {fsa.values.Dim(), 4};  // [num_rows, num_cols]
+  std::vector<int64_t> strides = {4, 1};               // in number of elements
+  auto options = torch::device(device).dtype(scalar_type);
+
+  auto tmp_scores = torch::from_blob(
+      fsa.values.Data(), sizes, strides,
+      [saved_region = fsa.values.GetRegion()](void *) {}, options);
+  return tmp_scores.index({"...", -1});
+}
+
+int32_t FsaClass::Properties() {
+  if (properties == 0) {
+    if (fsa.NumAxes() == 2) {
+      properties = GetFsaBasicProperties(fsa);
+    } else {
+      GetFsaVecBasicProperties(fsa, nullptr, &properties);
+    }
+    if (properties & kFsaPropertiesValid != kFsaPropertiesValid) {
+      K2_LOG(FATAL) << "Fsa is not valid, properties are : " << properties
+                    << " = " << PropertiesStr() << ", arcs are : " << fsa;
+    }
+  }
+  return properties;
+}
+
+std::string FsaClass::PropertiesStr() /*const*/ {
+  return FsaPropertiesAsString(Properties());
+}
+
+torch::Tensor FsaClass::Arcs() {
+  auto device = DeviceFromContext(fsa.Context());
+  auto scalar_type = caffe2::TypeMeta::Make<int32_t>();
+  // an Arc has 4 members
+  static_assert(sizeof(Arc) == 4 * sizeof(int32_t), "");
+
+  std::vector<int64_t> sizes = {fsa.values.Dim(), 4};  // [num_rows, num_cols]
+  std::vector<int64_t> strides = {4, 1};               // in number of elements
+  auto options = torch::device(device).dtype(scalar_type);
+
+  return torch::from_blob(
+      fsa.values.Data(), sizes, strides,
+      [saved_region = fsa.values.GetRegion()](void *) {}, options);
+}
+
+torch::Tensor FsaClass::Labels() /*const*/ { return Arcs().index({"...", 2}); }
+
+void FsaClass::SetLabels(torch::Tensor labels) {
+  K2_CHECK_EQ(labels.numel(), fsa.NumElements());
+  K2_CHECK_EQ(labels.scalar_type(), torch::kInt32);
+  Labels().copy_(labels);
+}
+
+FsaClass FsaClass::ToOtherContext(const ContextPtr &context) const {
```

Comment:
Also, I am not sure whether this function and the following `To` will be needed.
We can always construct an FSA on a given device and use it to construct a
`FsaClass`. After that, we don't need to move it to another device during decoding, I think.